### ISSUE TITLE
Better scoring system than 2 bonuses on every kill.

### ISSUE THREAD

**Issue by ashokae** [aichallenge/aichallenge#247]
Current scoring system of 2 points for any kill seem bit naive. Simply because one get 2 points even when ones own ant get dies. that is even when a collision of ants (which doesn't require much skill or strategy).

Ideally what we should measure is how good and effective, efficient a bot is in battle. So it's better to measure the kill-loss ratio instead. say, +1 for a kill and -1 for a loss. Then only the bots who are good in battle (ones who kill more ants than loosing) will get bonuses. Or may be we can give +2 for a kill and -1 for a loss to encourage bots to do battle.

One problem with the current scoring is that it doesn't take into account how many ants a bot is left with at the end of the game. For instance, towards end of a game if bot A has 70 ants, bot B has 85 and bot C has 10 and if all having the same score :

if bot C simply loose all it's ants (5 each for each opponent) at the end of the game while manging to kill only 6 ants of opponents, it will win the game! I have seen this actually happening in many games. reason is we don't give any credits for bots for remain with the larger army (being close to win the overall battle) at the end of the game. 

negative points for lost ants will address this.

**Comment by McLeopold** [comment]
The scoring needs to encourage a steady improvement in bot strength.  There are 2 things to consider: growth and fighting.  Currently you get 1 point per ant to reward efficient growth.  The -1 point per loss would counteract this and effectively be a bonus for the number of ants left at the end of the game.  Currently fighting well get's you 2 points per kill (shared among attackers.)  It is 2 points rather than one to emphasize that fighting is more important than growth.

In your scenario above, you say that the 3 bots all have equal points, but differing amounts of ants.  You'd have to consider how the situation came about.  The minimum score is 85 since B has 85 ants.  That means C has to have killed ants.  We could say it had spawned 25 ants and sent them into battle killing 25 for 50 points and losing the ants.  Those enemy ants had to come from somewhere, so you'd have to start balancing or figuring out who had what to make the numbers work.  At the end, you'd have to admit that C had been more active in spawning than it looks and had participated in more battles.  With the ending being only C initiating battles, and A and B not fighting each other even though the have more ants, I'm okay with the win going to C.

If you have specific examples of games where the scoring seems to favor the wrong bot, please post links here and I will look at them.

**Comment by ashokae** [comment]
Your arguments make sense and the example I have given may be not suitable to describe the issue I tried to convey. Will try to dig up those games I have seen which produced those odd results. Meanwhile the issue I tried to raise was :

We are giving bonuses only for kills but a kill can happen in many ways. 
1) a simple one on one exchange. (this is no indication that a bot is good at battle or willing to offer battle)
2) a bot is actually getting hammered in a battle but in the process manage to kill few of enemy ants as well. Say loosing 10 ants while killing 3, a total loss in a battle. This is simply a case of being poor in defense. 
3) a bot getting smart in it's battle logic and killing more enemy than it's own losses
4) a bot getting smart in deference and able to hold off an enemy attack by better deference. it kills 2 enemy ants for a loss of 2 of its own.

Now, only case 3 and 4 above shows signs of an improved bot. Giving bonuses for 1) and 2) can lead to odd results. For a two player game this wouldn't be a problem since when you getting 6 points for your 3 kills as in the example 2 above, your enemy is getting 20! so it is offset by enemy score. 

However, Just think about case 2 above in a 3 player game. Bot A which is winning is hammering bot C and making it go extinct. (killing all 10 of its ants while loosing only 3) but think of a Bot B in the same game which manges to avoid the carnage and loose only 2 ants while killing 2 of A ants. Here, Bot B has done nothing wrong, it is clearly better than Bot C in battle. but still bot C will go on and finish ahead of bot B in the game if the game is to stop at this stage. Since it collected 6 points to 4 points by bot B

In short, we are not rewarding skillful defensive play or skillful attacking play in battle. rewarding just the kills is not enough to measure how good a bot is really doing in battle. Looks like ants lost in battle is also be included somewhere in the equation to get a complete picture on a bots true ability in battles.

Will try to post few game links to show what is described above. Have seen many such games but unfortunately haven't recorded their links. If anybody have such links, please post.

**Comment by McLeopold** [comment]
I believe this is your argument: In general, a bot that participates in more battles will have a higher score, regardless of skill.  This means a less sophisticated bot may earn more points than a more sophisticated bot due to position on the board.

Hopefully the games are long enough that the differences in bots become apparent.  If a bot is truly more sophisticated, it shouldn't be in a position where the end of the game is near and a less sophisticated bot is even in score with in. 

A negative score for losing ants just counteracts the points for spawning an ant.  It reduces to giving 1 point for each surviving at at the end of the game, which we don't want.

I'd be willing to base ranking levels on some type of scale where bots are more likely to be given the same rank if the scores are in the same range:

```
Score | Rank
878  --> 1
800  --> 1
389  --> 2
99   --> 3
10   --> 3
```

Someone would have to come up with a good algorithm for that.  That would reduce the noise a little.

**Comment by McLeopold** [comment]
No further ideas, so I'm closing this.

**Comment by sir-macelon** [comment]
I have a proposal for an algorithm: https://gist.github.com/1231764

**Comment by ashokae** [comment]
I too have some further comments :

What about having the following as the scoring algorithm :

score = (no_of_ants_killed - number_of_ants_lost) \* battle_weight + Remaining_ants_at_the_end

Here, battle_weight is a number (say, 2 or 3). First part of the score will determine how good a bot is in battle and it will take both attack and defence skills. We could use the weight to increase/decrease the prominence to battles.

Second part acknowledges the final winner. In real world battles, normally the army who remains in the battle field with larger force is the natural victor. However, in our current formula this part is overlooked, hence we get upset results here and there with bots who have no remaining force winning the entire game just bcos it has traded off it's ants in last few turns before the static per map turn limit. Ants alive at the end of the battle has no say in the final out come, which is little odd in my mind. Above formula will address that.

Also, note that we don't give any bonus for spawning. I think that is how it should be spawning is mandatory for building an army and one who spawn more will benefit in the game, no particular reason to give a bonus for it. 

Another point is that it is possible for one to end up with a negative score. This too is justified as a negative score can happen only if someone is very weak in battles and has lost twice more ants than killed and end up with little or no army.  

Sorry if it is too late at this stage to come with ideas for changes like this. However, hope we could use the past accumulated game data to calculate possible results this new scoring algorithm would have produced and use those to decide. (without further extending the beta phase)

**Comment by McLeopold** [comment]
With a inconclusive score difference of 20%

```
Score | Rank
855  --> 1
690  --> 1
520  --> 1
```

690 is within 20% of 855 and 520 is within 20% of 690, but 520 is much lower than 855 yet they have the same rank.  We could find an example of this for any percentage.

Maybe if 2 bots are given the same rank, average the scores for the next check?

**Comment by ashokae** [comment]
Ok, if we average :

855 ->1
690 -> 1   (average of first 2 is 772)
620 ->1    (within 20% of 772) 

still, the difference between 620 and 850 is well above 20% but we are giving them same rank. If there is a 4 th bot 
as well......  

If the scores are inconclusive may be we can let the game run for few more turns ? If it is still too close after an extended turn limit, may be it's better to let whatever their close scores to determine the rank ? 

IMHO, when we arbitrarily try to define what is "inconclusive" and intervene, we might do more harm than good. (I may be wrong, we may need some testing to find out) So, rather than intervening, what if we do one or more of following :

1) increase turn limit in inconclusive looking scores in most cases they would become conclusive after few additional turns.
2) improve the scoring algorithm (suggested in a previous comment) to reduce upset results.
3) let the bots have little bit more info on their rank and deal with it. (sir_mecelons suggestion of sending an extra boolean value at each turn)

**Comment by sir-macelon** [comment]
What example would be for 10%? Basically this percentage shouldn't be as high as 20% in my opinion.

**Comment by McLeopold** [comment]
You'd probably need a nice clustering algo http://en.wikipedia.org/wiki/Cluster_analysis
